### PR TITLE
Pool Parameters (Contracts)

### PR DESCRIPTION
**Pull request by sebastianst**
Currently based on open PR #49 so must not be merged yet.

- contracts: Emit missing PoolDeployed event
- contracts: Add minimum stake amount and duration parameters

### CODE REVIEW

File: packages/contracts/contracts/YieldGate.sol [hunk 1]
```diff
@@ -68,13 +70,27 @@ contract YieldGate {
 }
 
 contract BeneficiaryPool {
-    event Staked(address indexed beneficiary, address indexed supporter, uint256 amount);
+    event Staked(
+        address indexed beneficiary,
+        address indexed supporter,
+        uint256 amount,
+        uint256 timeout
```
Comment (wottpal) [inline packages/contracts/contracts/YieldGate.sol:77]: Minor: I might find `timeout` not to be the perfect naming as this suggests that the staking ends when the time is over. But only the lock ends, right? But I also have no really good ideas beyond `lockTimeout` or `lockedUntil`.

File: packages/contracts/contracts/YieldGate.sol [hunk 2]
```diff
@@ -68,13 +70,27 @@ contract YieldGate {
 }
 
 contract BeneficiaryPool {
```
Comment (wottpal) [inline packages/contracts/contracts/YieldGate.sol:72]: Wanted to ask this for quite some time. Is it good practice to keep both contract definitions in the same file?

File: packages/contracts/contracts/YieldGate.sol [hunk 3]
```diff
@@ -93,23 +109,62 @@ contract BeneficiaryPool {
...
+    function emitParametersChanged(uint256 _minAmount, uint256 _minDuration) internal {
+        emit ParametersChanged(beneficiary, _minAmount, _minDuration);
     }
 
     // Stakes the sent ether on behalf of the provided supporter. The supporter
     // is usually msg.sender if staking on the transaction sender's behalf.
+    // The staking timeout is reset on each call, so prior stake is re-locked.
     function stake(address supporter) public payable {
         uint256 amount = msg.value;
+        require(amount > 0 && amount >= minAmount, "amount too low");
```
Comment (wottpal) [inline packages/contracts/contracts/YieldGate.sol:143]: Shouldn't we instead somehow check whether `supporters[supporter] += amount >= minAmount`. Because this way it's not possible to top-up donations with less than `minAmount`, right?

File: packages/contracts/contracts/YieldGate.sol [hunk 4]
```diff
@@ -93,23 +109,62 @@ contract BeneficiaryPool {
...
+
         supporters[supporter] += amount;
         totalStake += amount;
+        uint256 timeout = 0;
+        if (minDuration > 0) {
+            timeout = block.timestamp + minDuration;
```
Comment (wottpal) [inline packages/contracts/contracts/YieldGate.sol:149]: This also always resets the timeout when somebody tops-up, am I right? Which is kind of mean, isn't it 😆

File: packages/contracts/contracts/YieldGate.sol (hunk 1 above)
Comment (sebastianst) [reply packages/contracts/contracts/YieldGate.sol:77]: Agree, renamed this event var and also the contract storage mapping to `lockTimeouts`.

I'm just a fan of succinct names, but succinctness shouldn't come at the price of comprehensibility ;)

File: packages/contracts/contracts/YieldGate.sol (hunk 2 above)
Comment (sebastianst) [reply packages/contracts/contracts/YieldGate.sol:72]: I just did it in the context of the hackathon and then never moved it to a separate file... I'd say it's not bad practice if multiple contracts are closely related to keep them in the same file. But as complexity of individual contracts grows, they should probably be put into separate files. I can do it in this PR.

File: packages/contracts/contracts/YieldGate.sol (hunk 3 above)
Comment (sebastianst) [reply packages/contracts/contracts/YieldGate.sol:143]: We can also do that. Just wanted to keep things super simple in this iteration. But I think this is a fair change, so I'll do it.

File: packages/contracts/contracts/YieldGate.sol (hunk 4 above)
Comment (sebastianst) [reply packages/contracts/contracts/YieldGate.sol:149]: Yes it does reset on each top up, also added this fact to the function comment. I left it like that on purpose, because the alternative is to do more complex accounting of each top up, including storing each individual top up with lockTimeouts in an array etc... would get super messy and I want to keep things simple in this iteration.

Also, IMO this is actually a rather unimportant case to consider. I mean, if you top up, it seems like you want to support more, so why would you insist on getting the old stake out earlier? If that's the case, then better not top up at all and get your old stake out first ^^ I think in for first iteration it is sufficient to give the user a notice in the frontend that in case of a top up, the locked timeout is reset for their whole stake.

File: packages/contracts/contracts/YieldGate.sol (hunk 2 above)
Comment (wottpal) [reply packages/contracts/contracts/YieldGate.sol:72]: Ok, thanks for clarification. You decide whether to move or not :)

File: packages/contracts/contracts/YieldGate.sol (hunk 4 above)
Comment (wottpal) [reply packages/contracts/contracts/YieldGate.sol:149]: Agreed

**Comment by sebastianst** [comment]
@wottpal made me aware of an edge-case in which the logic should be extended: a reset of the min. duration to 0.
 
If the min. duration was > 0, a supporter stakes, and then the min. duration is reset to 0, if that user stakes again, the lock timeout is not reset and the user cannot unstake, either the old or new stake, until the old timeout has passed. This is not consistent because, if the min. duration was reset to a very low duration, but not 0, then staking any amount would reset the lock timeout using this new low duration.

I see two solutions:
1. Extend `stake` with an additional check
```sol
if (minDuration == 0 && lockTimeout[supporter] > now) {
    timeout = 0; lockTimeouts[supporter] = 0;
}
```
Note that only setting the timeout to 0 in case the current lock is in the future (instead of just checking > 0) is just a gas-optimization.

2. Extend `unstake` to skip the timeout check if `minDuration > 0`.

Solution 1 seems cleaner on first sight, as it makes sure lock timeouts are always set at staking time, which is more consistent. Also, there's an even edgier edge-case if only Solution 2 was implemented: if someone topups their stake while `minDuration` is 0, then the `minDuration` is increased again, that supporter would again not be able to access their stake, if an old lock is still set.

On the other hand, if only Solution 1 was implemented, any supporter could get their timeout reset anyways when the `minDuration` is 0 by just staking any (or the min.) amount and then immediately unstake. So to avoid forcing users to do this detour, it also makes sense to also implement Solution 2.

In short, it makes sense to do both extensions of the current logic.

**Comment by wottpal** [comment]
> In short, it makes sense to do both extensions of the current logic.
Makes sense and I agree. Thanks for this detailed explanation.

**Comment by sebastianst** [comment]
Just realized that a simpler version of S. 1 is to always set the `lockTimeouts[supporter]` storage to `0` if `minDuration == 0` and to `now+minDuration` otherwise, because resetting a storage slot from 0 to 0 just costs as much as loading it (SLOAD), see https://eips.ethereum.org/EIPS/eip-2200 for more info. Will do it like that because simpler logic is always better :)

**Review by wottpal — APPROVED** [review]
